File: LinkedList/concepts/reverse_ll.cpp
#include<iostream>
using namespace std;

class Node{
    public:
        int data;
        Node* next;
    
        Node(int data){
            this->data = data;
            this->next = NULL;
        }

        Node *reverse(Node* head){
            Node* curr = head;
            Node* prev;
            Node* after;
            prev->next=NULL;
            prev->data=curr->data;
            while(curr->next!=NULL){
                after = curr->next;
                curr=curr->next;
                curr->next=prev;
                prev = curr;
                curr = after;
            }
            head = prev;
            return head;
        }

        void traverse_LL(Node* head){
            Node * curr = head;
            cout<<"head : ";
            while(curr->next!=NULL){
                cout<<curr->data<<"->";
                curr=curr->next;
            }
            cout<<"NULL.";
        }

        

};

int main() {
    
    Node* head = new Node(10);
    Node* one = new Node(11);
    head->next = one;
    Node* two = new Node(12);
    one->next = two;
    Node* three = new Node(13);
    two->next = three;
    // Linked List : 10->11->12->13->NULL

    head->traverse_LL(head);


    return 0;
}

File: LinkedList/reverseLL.py

class Node():
    def __init__(self,data,next=None):
        self.data = data
        self.next = next
        
        
class LinkedList():
    
    def createLL(self):
        arr = [10,20,30,40,50]
        head = Node(10)
        curr = head
        for i in range(1,len(arr)):
            temp = Node(arr[i])
            curr.next = temp
            curr = temp
            
        return head
            
    def reverseLL(self,head):
        """
        head = 10->20->30->40->50->None
        """
        curr = head
        prev = None
        
        
        
        while(curr!=None):
            next = curr.next
            curr.next = prev
            prev = curr
            curr = next

        return prev
            
        
    def traverseLL(self,head):
        curr = head
        while(curr!=None):
            print(f"{curr.data}->")
            curr = curr.next


if __name__ == "__main__":
    
    ll = LinkedList()
    head = ll.createLL()
    print(f"LinkedList befor reverse:")
    ll.traverseLL(head)
    print(f"\nLinked list after reverse")
    new_head = ll.reverseLL(head)
    ll.traverseLL(new_head)
    
    
                
        